### PR TITLE
Make UI for restricted scenery, and make restricted paths and terrain

### PR DESCRIPTION
**Pull request by MaxMallon**
I've added created UI for restricted scenery, and added the possibility to restrict paths and terrain.
image

In the object selection screen, any object types that support restriction will now show a column 'Restricted'. Any item that is selected shows a checkmark in that column. 
I've also added a filter option to filter on restricted only. 

All types of scenery, but not scenery groups, legacy paths, path surfaces, path railings, terrain surfaces and terrain edges are supported. 
Any objects listed as restricted do not appear during scenario play. Turning on sandbox mode makes them appear again. 

The "Road" path type uses a separate flag which as far as I can tell is inherit to it specifically. The flag basically does the same, but is linked with the actual object, and not with the current object settings for a given scenario.

### CODE REVIEW

**Comment by oli414** [comment]
Is the screenshot still up-to-date with how it currently looks after your changes?

**Review by oli414 — COMMENTED** [review]
I'm wondering if it might be better to have a single list of restricted objects, turning the _restrictedScenery list into something more generic.

**Comment by MaxMallon** [comment]
> Is the screenshot still up-to-date with how it currently looks after your changes?

Yes, it is



> I'm wondering if it might be better to have a single list of restricted objects, turning the _restrictedScenery list into something more generic.

Possibly? I just went with what was already there. I'm not sure what sort of more generic type would cover all of them though.

File: src/openrct2-ui/windows/EditorObjectSelection.cpp [hunk 1]
```diff
@@ -584,21 +600,35 @@ class EditorObjectSelectionWindow final : public Window
         }
 
         uint32_t inputFlags = INPUT_FLAG_EDITOR_OBJECT_1 | INPUT_FLAG_EDITOR_OBJECT_SELECT_OBJECTS_IN_SCENERY_GROUP;
-        // If already selected
-        if (!(object_selection_flags & ObjectSelectionFlags::Selected))
-            inputFlags |= INPUT_FLAG_EDITOR_OBJECT_SELECT;
 
-        _gSceneryGroupPartialSelectError = false;
-        if (!window_editor_object_selection_select_object(0, inputFlags, listItem->repositoryItem))
+        if (screenCoords.x < widgets[WIDX_LIST].width() - 25
```
[the remark this answers is not in the document]
Comment (MaxMallon) [reply src/openrct2-ui/windows/EditorObjectSelection.cpp]: Fairies and unicorns have been eradicated now

**Comment by Gymnasiast** [comment]
> I'm wondering if it might be better to have a single list of restricted objects, turning the _restrictedScenery list into something more generic.

I was thinking that putting checkboxes in the Theming window to toggle this might be an idea too. That will also make it quite clear which scenery will be hidden when you’re decorating your park in the Scenario Editor (or in sandbox mode).

**Comment by MaxMallon** [comment]
> > I'm wondering if it might be better to have a single list of restricted objects, turning the _restrictedScenery list into something more generic.
> 
> I was thinking that putting checkboxes in the Theming window to toggle this might be an idea too. That will also make it quite clear which scenery will be hidden when you’re decorating your park in the Scenario Editor (or in sandbox mode).

That might work well for scenery, but what about all the other types of objects that could be locked like this? Like paths etc, that I also addressed in this PR?

**Comment by Gymnasiast** [comment]
I don’t think those should be locked to begin with.

Maybe other team members want to chime in on this?

**Comment by IntelOrca** [comment]
I only remember implementing it for scenery items, and simply for the purpose of more accurate RCT 1 import.
I think RCT 1 restricted you from building roads, so technically adding a restriction for roads would more faithful to RCT 1. Not sure about rides, I suppose you could have a scenario where you have pre-existing rides which you can not build any new instances.

**Comment by karst** [comment]
If memory serves right, roads weren't available outside scenario editor in RCT2. I could imagine wanting to make a neighborhood outside park limits but not have these houses available, or a mining area where old huts and track wouldn't be available to plop down in the scenario itself. This could definitely be beneficial to making scenarios more interesting.

**Comment by MaxMallon** [comment]
Or even things like a unique geological feature -terrain type- inside the park, that if you destroy it, you can't restore.

**Comment by Gymnasiast** [comment]
Roads are already restricted to the Scenario Editor - that flag was in vanilla.

**Comment by AaronVanGeffen** [comment]
Personally, I think the added checkboxes clutter the window quite a bit. I think we should consider hiding the restricted column by default, making it opt-in to those who want or need the extra control.

**Comment by karst** [comment]
I personally think it looks fine to lock it behind sandbox mode only, which is what is implemented here if I read correctly.

**Comment by MASHzj** [comment]
I like the view with the "restricted" checkboxes column. To address the worry about cluttering, it may be desirable to hide this column behind the "Advanced"-button?
This PR would fix some issues described in #17610. If a player wants to get access to hidden scenery in RCT1 scenarios, they can uncheck those checkboxes. And it would provide a much easier workaround to get access to jumping fountains in scenarios where it's a single researchable item. Jumping fountains are often not even accessible under sandbox mode in those scenarios.

**Comment by MASHzj** [comment]
@MaxMallon Are you still willing to work on this?

**Comment by karst** [comment]
Keep this open please

**Pull request closed without merging**